viz/src/bin/draw_occupancy_grid.rs

```diff
@@ -23,85 +25,125 @@ fn draw_frame_marker(window: &mut kiss3d::window::Window, pose: &kuba::Pose3, le
     );
 }
 
+struct GridUpdater {
+    grid_map: std::sync::Mutex<kuba::LidarOccupancyGridMap3>,
+    changed_cells: std::sync::Mutex<Vec<kuba::Cell<na::U3>>>,
+    finished: atomic::AtomicBool,
+}
+
+impl GridUpdater {
+    fn new() -> GridUpdater {
+        GridUpdater {
+            grid_map: std::sync::Mutex::new(kuba::LidarOccupancyGridMap3::default()),
+            changed_cells: std::sync::Mutex::new(vec![]),
+            finished: atomic::AtomicBool::new(true),
+        }
+    }
+}
+
 struct AppState {
     poses: Vec<kuba::Pose3>,
     point_clouds: Vec<kuba::PointCloud3>,
     index: usize,
-    loop_counter: usize,
-    first_loop: bool,
-    grid_map: kuba::LidarOccupancyGridMap3,
+    grid_updater: std::sync::Arc<GridUpdater>,
+    tracked_cells: std::collections::HashMap<kuba::Cell<na::U3>, na::Point3<f32>>,
```

Comment:
This could just be a HashSet of cells. Then the point conversion would happen while iterating over them.